### PR TITLE
smartconract: generate RPC binding wrappers for events

### LINKED ISSUE #2891: Autogenerate structures for events and stack item -> struct conversion code
See nspcc-dev/neofs-contract#305. Most of the time we have simple types there, so it should work even for non-Go contracts right away.

Consider some utility functions to deal with application logs and events directly as well.

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Close #2891.

TODO:

- [x] Handle all event usages when `--guess-eventtypes` is on, not only the first one.
- [x] Prettify event names before saving bindings configuration, i.e. at the compile time.
- [x] Properly handle imports when using user-defined types configuration.
- [x] Add tests for `--guess-eventtypes` option disabled/enabled with complicated and nested event types.

### CODE REVIEW

**Comment by AnnaShaleva** [comment]
@roman-khimov, here's the draft PR and it contains a lot of TODOs, but the main idea is implemented as discussed and I'd like you to check whether it's something we'd like to have (from the user side).

Note on the commits structure: its far from being perfect and I know it. The earlier commits make it possible to fetch event structure from the contract code. The last commit makes this behaviour a part of extension and allows the user to specify the events structure manually in the config, e.g. like this (pay attention to the `extendedtype` and `namedtypes` sections):
```
name: "HelloWorld contract"
safemethods: ["oneSafeMethod"]
events:
  - name: main! ev
    parameters:
      - name: int
        type: Integer
      - name: bool
        type: Boolean
      - name: string
        type: String
  - name: transfer funds
    parameters:
      - name: from   whom
        type: Hash160
      - name: to
        type: Any
      - name: structArg
        type: Struct
        extendedtype:
          base: Array # This one can safely be removed, as can be retrieved from the event parameter. And need to be.
          name: main.S
  - name: map notification
    parameters:
      - name: the map itself
        type: Map
namedtypes:
  main.S:
    base: Array
    name: main.S
    fields:
      - field: I
        base: Integer
      - field: B
        base: Boolean
supportedstandards: []
```

So if we're OK with the result this PR produces and the way how it is performed, then I'll reorganise the commits structure and complete all TODOs hidden in the code.

File: pkg/compiler/compiler.go [hunk 1]
```diff
@@ -263,7 +263,7 @@ func CompileAndSave(src string, o *Options) ([]byte, error) {
 			for j, p := range e.Parameters {
 				params[j] = DebugParam{
 					Name: p.Name,
-					Type: p.Type.String(),
+					Type: p.Type.ConvertToStackitemType().String(),
```
Comment (AnnaShaleva) [inline pkg/compiler/compiler.go:295]: And BTW, this is a bag and probably this fix needs to be added as a separate PR. @roman-khimov, do we need it?

File: pkg/smartcontract/binding/generate.go [hunk 2]
```diff
@@ -48,14 +48,23 @@ const Hash = "{{ .Hash }}"
...
+		// NamedTypes contains exported structured types that have some name (even
+		// if the original structure doesn't) and a number of internal fields. The
+		// map key is in the form of `namespace.name`, the value is fully-qualified
+		// and possibly nested description of the type structure.
+		NamedTypes map[string]ExtendedType `yaml:"namedtypes,omitempty"`
+		// Types contains type structure description for various types used in
+		// smartcontract. The map key has one of the following forms:
+		// - `methodName` for method return value;
+		// - `mathodName.paramName` for method's parameter value;
+		// - `eventName.paramName` for event's parameter value.
```
Comment (roman-khimov) [inline pkg/smartcontract/binding/generate.go:65]: I don't see any functonal changes in this hunk, so it better be done separately from this commit.

File: pkg/smartcontract/rpcbinding/binding.go [hunk 3]
```diff
@@ -576,6 +666,50 @@ func scTemplateToRPC(cfg binding.Config, ctr ContractTmpl, imports map[string]st
 	if len(cfg.NamedTypes) > 0 {
 		imports["errors"] = struct{}{}
 	}
+	for _, abiEvent := range cfg.Manifest.ABI.Events {
+		eTmp := CustomEventTemplate{
+			// TODO: proper event name is better to be set right into config binding in normal form.
```
Comment (roman-khimov) [inline pkg/smartcontract/rpcbinding/binding.go:671]: You can have anything in the manifest.

File: pkg/compiler/debug.go [hunk 4]
```diff
@@ -110,11 +110,13 @@ type DebugRange struct {
...
-	Type         string                  `json:"type"`
-	RealType     binding.Override        `json:"-"`
-	ExtendedType *binding.ExtendedType   `json:"-"`
-	TypeSC       smartcontract.ParamType `json:"-"`
+	Name string `json:"name"`
+	// Type is a string representation of stackitem.Type of the variable (VM's type).
+	Type         string                `json:"type"`
+	RealType     binding.Override      `json:"-"`
+	ExtendedType *binding.ExtendedType `json:"-"`
+	// TypeSC is a smartcontract.Type of the variable (SmartContract's type).
```
Comment (roman-khimov) [inline pkg/compiler/debug.go:118]: No functional changes -> something different.

File: pkg/compiler/compiler.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/compiler/compiler.go:295]: Hmm. https://github.com/neo-project/proposals/blob/master/nep-19.mediawiki#typename

**Comment by AnnaShaleva** [comment]
Not ready for review yet.

File: pkg/compiler/compiler.go (hunk 1 above)
Comment (AnnaShaleva) [reply pkg/compiler/compiler.go]: We have several types mismatch in our code and a lot of them in our tests, I'll create a separate PR to fix that. I alsoo need to re-check how debugger works with it.

File: pkg/smartcontract/rpcbinding/binding.go [hunk 5]
```diff
@@ -399,7 +399,12 @@ func extendedTypeToGo(et binding.ExtendedType, named map[string]binding.Extended
 
 	case smartcontract.MapType:
 		kt, _ := extendedTypeToGo(binding.ExtendedType{Base: et.Key}, named)
-		vt, _ := extendedTypeToGo(*et.Value, named)
+		var vt string
+		if et.Value != nil {
+			vt, _ = extendedTypeToGo(*et.Value, named)
+		} else {
+			vt = "any"
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/smartcontract/rpcbinding/binding.go:519]: Not fixed yet, and currently I'm not sure that this code is needed. I'll recheck and write a test.

File: cli/smartcontract/generate_test.go [hunk 6]
```diff
@@ -316,11 +316,11 @@ func NewReader(invoker Invoker) *ContractReader {
 	return &ContractReader{invoker}
 }
 
-
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply cli/smartcontract/generate_test.go:319]: Fixed, now we have a new rule for writing blocks of code to our template: new line before the block starts and new line after the block ends. Very native, it fixes all the indents.

File: cli/smartcontract/generate_test.go [hunk 7]
```diff
@@ -316,11 +316,11 @@ func NewReader(invoker Invoker) *ContractReader {
...
 // Get invokes `+"`get`"+` method of contract.
 func (c *ContractReader) Get() (*big.Int, error) {
 	return unwrap.BigInt(c.invoker.Call(Hash, "get"))
 }
+
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply cli/smartcontract/generate_test.go]: Fixed.

File: cli/smartcontract/testdata/gas/gas.go [hunk 8]
```diff
@@ -2,13 +2,25 @@
...
 	"github.com/nspcc-dev/neo-go/pkg/rpcclient/nep17"
 	"github.com/nspcc-dev/neo-go/pkg/util"
+	"github.com/nspcc-dev/neo-go/pkg/vm/stackitem"
+	"math/big"
 )
 
 // Hash contains contract hash.
 var Hash = util.Uint160{0xcf, 0x76, 0xe2, 0x8b, 0xd0, 0x6, 0x2c, 0x4a, 0x47, 0x8e, 0xe3, 0x55, 0x61, 0x1, 0x13, 0x19, 0xf3, 0xcf, 0xa4, 0xd2}
 
+// TransferEvent represents event emitted by the contract.
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply cli/smartcontract/testdata/gas/gas.go]: Fixed.

File: cli/smartcontract/testdata/gas/gas.go [hunk 9]
```diff
@@ -45,3 +57,83 @@ func New(actor Actor) *Contract {
 	return &Contract{ContractReader{nep17t.TokenReader, actor}, nep17t.TokenWriter, actor}
 }
 
+// TransferEventFromApplicationLog retrieves TransferEvent from the
+// provided ApplicationLog located at the specified index in the events list
+// of the specified execution.
+func TransferEventFromApplicationLog(log *result.ApplicationLog, executionIdx, eventIdx int) (*TransferEvent, error) {
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply cli/smartcontract/testdata/gas/gas.go]: Agree, fixed.

File: cli/smartcontract/testdata/nameservice/nns.go [hunk 10]
```diff
@@ -16,6 +18,28 @@ import (
 // Hash contains contract hash.
 var Hash = util.Uint160{0xde, 0x46, 0x5f, 0x5d, 0x50, 0x57, 0xcf, 0x33, 0x28, 0x47, 0x94, 0xc5, 0xcf, 0xc2, 0xc, 0x69, 0x37, 0x1c, 0xac, 0x50}
 
+// TransferEvent represents event emitted by the contract.
+type TransferEvent struct {
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply cli/smartcontract/testdata/nameservice/nns.go]: Fixed, now generator ignores standard events. I've also manually added event deserialization helpers to `nep11` and `nep17`, is it OK?

File: cli/smartcontract/testdata/nex/nex.go [hunk 11]
```diff
@@ -2,17 +2,36 @@
...
 	"github.com/nspcc-dev/neo-go/pkg/rpcclient/unwrap"
 	"github.com/nspcc-dev/neo-go/pkg/util"
+	"github.com/nspcc-dev/neo-go/pkg/vm/stackitem"
 	"math/big"
 )
 
 // Hash contains contract hash.
 var Hash = util.Uint160{0xa8, 0x1a, 0xa1, 0xf0, 0x4b, 0xf, 0xdc, 0x4a, 0xa2, 0xce, 0xd5, 0xbf, 0xc6, 0x22, 0xcf, 0xe8, 0x9, 0x7f, 0xa6, 0xa2}
 
+// TransferEvent represents event emitted by the contract.
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply cli/smartcontract/testdata/nex/nex.go]: Fixed.

File: pkg/smartcontract/binding/generate.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/smartcontract/binding/generate.go:65]: No functional changes, just a comment that clarify things a bit. Fixed.

Comment (AnnaShaleva) [reply pkg/smartcontract/binding/generate.go:65]: Fixed.

File: pkg/smartcontract/rpcbinding/binding.go (hunk 3 above)
Comment (AnnaShaleva) [reply pkg/smartcontract/rpcbinding/binding.go]: Fixed in a separate commit.

File: pkg/compiler/debug.go (hunk 4 above)
Comment (AnnaShaleva) [reply pkg/compiler/debug.go]: Moved to a separate PR.

File: pkg/compiler/compiler.go [hunk 12]
```diff
@@ -309,36 +338,44 @@ func CompileAndSave(src string, o *Options) ([]byte, error) {
...
-					// non-event's parameter name.
-					if p.RealType.TypeName != "" {
-						cfg.Overrides[pname] = p.RealType
+			}
+		}
+		if o.GuessEventTypes {
+			if len(di.EmittedEvents) > 0 {
+				for eventName, eventUsages := range di.EmittedEvents {
+					// Take into account the first usage only.
+					// TODO: extend it to the rest of invocations.
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/compiler/compiler.go:361]: I've firstly added the whole set of extended types that are used in `runtime.Notify` calls for this notification to the bindings config. But then we need to generate several event structures for a single event (if params differs from invocation to invocation). This requires a separate config section into our bindings config file. And after that I'm wondering, do we ever need this?

Another way I was considering is to check that all the invocations has the same parameter types, and if not, then write a warning to the output and generate a single wrapper for the first usage. Not pretty solution, but can exist as far.

**Comment by AnnaShaleva** [comment]
@roman-khimov, I still need to write a set of missing tests, but the main flow of generation is there and can be reviewed one more time.

File: cli/smartcontract/testdata/nameservice/nns.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (roman-khimov) [reply cli/smartcontract/testdata/nameservice/nns.go]: Sure.

File: pkg/compiler/compiler.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/compiler/compiler.go:361]: > if params differs from invocation to invocation

Then it's either `Any` or a bug (more likely). We don't need same events with the same names but different arguments.

File: cli/smartcontract/generate_test.go [hunk 13]
```diff
@@ -325,6 +325,8 @@ func (c *ContractReader) Get() (*big.Int, error) {
 }
 
 func TestGenerateRPCBindings(t *testing.T) {
+	const rewriteExpectedOutputs = false
```
Comment (roman-khimov) [inline cli/smartcontract/generate_test.go:327]: Make it global? You likely want to update all of them at once, although I've done it manually previously (one liner).

File: pkg/smartcontract/rpcbinding/binding.go [hunk 14]
```diff
@@ -162,9 +176,9 @@ type ContractReader struct {
 	{{end -}}
 	invoker Invoker
 }
-
```
Comment (roman-khimov) [inline pkg/smartcontract/rpcbinding/binding.go:165]: Mixing spacing changes and logic ones in one commit leaves some space for improvement.

File: pkg/rpcclient/nep11/base.go [hunk 15]
```diff
@@ -246,3 +247,71 @@ func UnwrapKnownProperties(m *stackitem.Map, err error) (map[string]string, erro
 	}
 	return res, nil
 }
+
+// TransferEventsFromApplicationLog retrieves all emitted TransferEvents from the
+// provided ApplicationLog.
+func TransferEventsFromApplicationLog(log *result.ApplicationLog) ([]*TransferEvent, error) {
```
Comment (roman-khimov) [inline pkg/rpcclient/nep11/base.go:253]: @notimetoname would probably be interested to see this and imagine what it may look like if we're to use generics.

File: pkg/smartcontract/rpcbinding/binding.go (hunk 14 above)
Comment (AnnaShaleva) [reply pkg/smartcontract/rpcbinding/binding.go:165]: Agree, I moved a part which fixes the indentation to a separate commit.

File: cli/smartcontract/generate_test.go (hunk 13 above)
Comment (AnnaShaleva) [reply cli/smartcontract/generate_test.go]: Fixed.

File: pkg/rpcclient/nep11/base.go (hunk 15 above)
Comment (AnnaShaleva) [reply pkg/rpcclient/nep11/base.go:253]: We can make use of generics here as far, https://github.com/nspcc-dev/neo-go/issues/2377#issuecomment-1563185046.

File: pkg/compiler/compiler.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/compiler/compiler.go]: I've added a check in a separate commit, but I need to test it.

**Comment by AnnaShaleva** [comment]
@roman-khimov, there's still a space for improvements, new issues are coming while I'm testing the implemented functionality. There'are still a couple of TODOs left, but I beleive that the most significant ones are resolved and can be reviewed.

**Comment by AnnaShaleva** [comment]
@roman-khimov, I've finished the TODOs and tests for main functionality, this PR needs you!

**Comment by roman-khimov** [comment]
> this PR needs you

18 commits. +5,207 −321. Yeah, thanks!

File: cli/smartcontract/smart_contract.go [hunk 16]
```diff
@@ -171,6 +172,10 @@ func NewCommands() []cli.Command {
 						Name:  "no-permissions",
 						Usage: "do not check if invoked contracts are allowed in manifest",
 					},
+					cli.BoolFlag{
+						Name:  "guess-eventtypes",
```
Comment (roman-khimov) [inline cli/smartcontract/smart_contract.go:176]: `docs/compiler.md`?

File: pkg/compiler/compiler.go [hunk 17]
```diff
@@ -78,6 +90,23 @@ type Options struct {
...
+	Name       string            `json:"name"`
+	Parameters []HybridParameter `json:"parameters"`
+}
+
+// HybridParameter contains the manifest's event parameter description united with
+// the extended type description for this parameter. It is applied for the smart
+// contract configuration file only.
+type HybridParameter struct {
+	manifest.Parameter `yaml:",inline"`
+	ExtendedType       *binding.ExtendedType `yaml:"extendedtype,omitempty"`
```
Comment (roman-khimov) [inline pkg/compiler/compiler.go:108]: `docs/compiler.md`?

File: pkg/rpcclient/nep11/base.go [hunk 18]
```diff
@@ -246,3 +247,71 @@ func UnwrapKnownProperties(m *stackitem.Map, err error) (map[string]string, erro
 	}
 	return res, nil
 }
+
+// TransferEventsFromApplicationLog retrieves all emitted TransferEvents from the
+// provided ApplicationLog.
```
Comment (roman-khimov) [inline pkg/rpcclient/nep11/base.go:252]: `[result.ApplicationLog]` to have a doc link.

File: pkg/rpcclient/nep11/base.go [hunk 19]
```diff
@@ -246,3 +247,71 @@ func UnwrapKnownProperties(m *stackitem.Map, err error) (map[string]string, erro
...
+			if err != nil {
+				return nil, fmt.Errorf("failed to decode event from stackitem (event #%d, execution #%d): %w", j, i, err)
+			}
+			res = append(res, event)
+		}
+	}
+	return res, nil
+}
+
+// FromStackItem converts provided stackitem.Array to TransferEvent and
```
Comment (roman-khimov) [inline pkg/rpcclient/nep11/base.go:274]: `[stackitem.Array]`

File: pkg/rpcclient/nep11/base.go [hunk 20]
```diff
@@ -246,3 +247,71 @@ func UnwrapKnownProperties(m *stackitem.Map, err error) (map[string]string, erro
...
+// returns an error if so.
```
Comment (roman-khimov) [inline pkg/rpcclient/nep11/base.go:275]: So it always returns an error?

File: pkg/rpcclient/nep17/nep17.go [hunk 21]
```diff
@@ -147,3 +150,66 @@ func (t *TokenWriter) MultiTransferUnsigned(params []TransferParameters) (*trans
 	}
 	return t.actor.MakeUnsignedRun(script, nil)
 }
+
+// TransferEventsFromApplicationLog retrieves all emitted TransferEvents from the
+// provided ApplicationLog.
```
Comment (roman-khimov) [inline pkg/rpcclient/nep17/nep17.go:155]: Ditto.

File: pkg/rpcclient/nep17/nep17.go [hunk 22]
```diff
@@ -147,3 +150,66 @@ func (t *TokenWriter) MultiTransferUnsigned(params []TransferParameters) (*trans
...
+				return nil, fmt.Errorf("failed to decode event from stackitem (event #%d, execution #%d): %w", j, i, err)
+			}
+			res = append(res, event)
+		}
+	}
+	return res, nil
+}
+
+// FromStackItem converts provided stackitem.Array to TransferEvent and
+// returns an error if so.
```
Comment (roman-khimov) [inline pkg/rpcclient/nep17/nep17.go:178]: And here.

File: pkg/smartcontract/rpcbinding/binding.go [hunk 23]
```diff
@@ -675,36 +674,31 @@ func scTemplateToRPC(cfg binding.Config, ctr ContractTmpl, imports map[string]st
...
+			Name:         eBindingName,
 			ManifestName: abiEvent.Name,
 		}
-		var varnames = make(map[string]bool)
 		for i := range abiEvent.Parameters {
-			name := abiEvent.Parameters[i].Name
-			fullPName := abiEvent.Name + "." + name
+			pBindingName := ToParameterBindingName(abiEvent.Parameters[i].Name)
+			fullPName := eBindingName + "." + pBindingName
 			typeStr, pkg := scTypeConverter(fullPName, abiEvent.Parameters[i].Type, &cfg)
```
Comment (roman-khimov) [inline pkg/smartcontract/rpcbinding/binding.go:699]: Suppose we have a new NEP with extended type data. Will this scheme still work with compilers pushing random things into names?

File: cli/smartcontract/generate_test.go [hunk 24]
```diff
@@ -499,3 +499,52 @@ callflags:
...
+	})
+	t.Run("invalid number of params", func(t *testing.T) {
+		check(t, filepath.Join("testdata", "invalid6"), "inconsistent usages of event `SomeEvent` against config: number of params mismatch: 2 vs 1")
+	})
+	/*
+		// TODO: this on is a controversial one. If event information is provided in the config file, then conversion code
+		// will be emitted by the compiler according to the parameter type provided via config. Thus, we can be sure that
+		// either event parameter has the type specified in the config file or the execution of the contract will fail.
+		// Thus, this testcase is always failing (no compilation error occures).
+		// Question: do we want to compare `RealType` of the emitted parameter with the one expected in the manifest?
```
Comment (roman-khimov) [inline cli/smartcontract/generate_test.go:542]: I think this failed to compile previously, `"s", 5` can't be the same as `"s", "s"` and if it can be checked then a compilation error is expected.

File: cli/smartcontract/testdata/invalid9/invalid.yml [hunk 25]
```diff
@@ -0,0 +1,16 @@
...
+events:
+  - name: SomeEvent
+    parameters:
+      - name: p1
+        type: Array
+        extendedtype:
+          base: Array
+          name: invalid9.NamedStruct
+namedtypes:
+  invalid9.NamedStruct:
```
Comment (roman-khimov) [inline cli/smartcontract/testdata/invalid9/invalid.yml:11]: But it's OK to have a `NamedStruct` here?

Comment (AnnaShaleva) [reply cli/smartcontract/testdata/invalid9/invalid.yml:11]: Yes.

File: pkg/smartcontract/rpcbinding/binding.go (hunk 23 above)
Comment (AnnaShaleva) [reply pkg/smartcontract/rpcbinding/binding.go:699]: > compilers pushing random things into names

It's just *our* compiler that pushes prettified names into config *that may be used only by our compiler*. The parameter names must be prettified to generate proper go code (prettification includes capitalization, removing spaces and all not-a-go-name-ish garbage that user may put into event/parameter name. This step is needed anyway and, as discussed in https://github.com/nspcc-dev/neo-go/pull/3008#discussion_r1191228640, the config should contain already prettified names, not just the user's input.

> Suppose we have a new NEP with extended type data

This scheme will work with the new NEP, and I don't expect that this part of code will be removed or rewrited, because notification names can still contain any garbage as far as notification parameter names, and the new NEP doesn't solve this problem. I.e. we need to prettify them anyway.

File: pkg/rpcclient/nep17/nep17.go (hunk 22 above)
Comment (AnnaShaleva) [reply pkg/rpcclient/nep17/nep17.go]: Fixed.

File: pkg/rpcclient/nep17/nep17.go (hunk 21 above)
Comment (AnnaShaleva) [reply pkg/rpcclient/nep17/nep17.go]: Fixed.

File: pkg/rpcclient/nep11/base.go (hunk 20 above)
Comment (AnnaShaleva) [reply pkg/rpcclient/nep11/base.go]: No, fixed.

File: pkg/rpcclient/nep11/base.go (hunk 19 above)
Comment (AnnaShaleva) [reply pkg/rpcclient/nep11/base.go]: Fixed.

File: pkg/rpcclient/nep11/base.go (hunk 18 above)
Comment (AnnaShaleva) [reply pkg/rpcclient/nep11/base.go]: Fixed.

File: pkg/compiler/compiler.go (hunk 17 above)
Comment (AnnaShaleva) [reply pkg/compiler/compiler.go:108]: Fixed.

File: pkg/compiler/compiler.go [hunk 26]
```diff
@@ -309,6 +309,38 @@ func CompileAndSave(src string, o *Options) ([]byte, error) {
 		if len(di.NamedTypes) > 0 {
 			cfg.NamedTypes = di.NamedTypes
 		}
+		if len(di.EmittedEvents) > 0 {
+			for eventName, eventUsages := range di.EmittedEvents {
+				// QUESTION: (RESOLVED, need to make an extension, e.g --guess-event-type-on-compile-time flag to compilation)
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/compiler/compiler.go]: Removed.

File: cli/smartcontract/generate_test.go (hunk 24 above)
Comment (roman-khimov) [reply cli/smartcontract/generate_test.go:542]: You'll get binary `5` in the `ByteString` instead of (likely expected) the string "5".

File: pkg/smartcontract/rpcbinding/binding.go [hunk 27]
```diff
@@ -248,10 +246,15 @@ func (res *{{toTypeName $name}}) FromStackItem(item stackitem.Item) error {
...
-{{end}}
+{{- end}}
 	return nil
 }
-{{end}}`
+{{end -}}`
+
+	srcTmpl = safemethodDefinition +
+		methodDefinition +
+		bindingDefinition
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/smartcontract/rpcbinding/binding.go]: It doesn't matter, because `methodDefinition` and `safemethodDefinition` are just templates, they won't be added into the resulting code.

But let's have the binding firstly, fixed.

File: pkg/smartcontract/rpcbinding/binding.go [hunk 28]
```diff
@@ -221,25 +221,35 @@ func itemTo{{toTypeName $name}}(item stackitem.Item, err error) (*{{toTypeName $
 	if err != nil {
 		return nil, err
 	}
+	var res = new({{toTypeName $name}})
+	err = res.FromStackItem(item)
+	return res, err
+}
+
+// FromStackItem retrieves fields of {{toTypeName $name}} from the given stack item
+// and returns an error if so.
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/smartcontract/rpcbinding/binding.go]: Fixed.

[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/smartcontract/rpcbinding/binding.go]: (in a separate commit)

File: cli/smartcontract/generate_test.go (hunk 24 above)
Comment (AnnaShaleva) [reply cli/smartcontract/generate_test.go:542]: I'll remove this comment for now, let's have a test for it commented out.

File: pkg/smartcontract/rpcbinding/binding.go (hunk 23 above)
Comment (AnnaShaleva) [reply pkg/smartcontract/rpcbinding/binding.go:699]: Yes, we do! Here's the test: `TestAssistedRPCBindings`, check the `checkBinding(filepath.Join("testdata", "notifications"), true, "_guessed")` and the resulting output located in the `neo-go/cli/smartcontract/testdata/notifications/rpcbindings_guessed.out`. We also have the same result if no `--guess-eventtypes` is used and no additional information is specified in the contract config. (see the `! complicated name %$#` notification in the contract config)

**Review by roman-khimov — APPROVED** [review]
